pip-audit silently audits only 2 of 4 requirement files, and the job stays green

## Symptom

The `security-scan` job reports success while the dependency audit has in fact
failed and two of the four requirement files were never audited.

In [this run](https://github.com/nuetzliches/powerbrain/actions/runs/32072560871/job/95518774862#logs)
the step log reads:

```
No known vulnerabilities found                     <- mcp-server/requirements.txt
Found 3 known vulnerabilities in 1 package         <- ingestion/requirements.txt
##[error]Process completed with exit code 1
```

There is no output for `pb-proxy/requirements.txt` or `worker/requirements.txt`.

## Cause

Two things combine, in `.github/workflows/pr-validate.yml`:

```yaml
- name: Dependency audit (pip-audit)
  continue-on-error: true
  run: |
    pip-audit -r mcp-server/requirements.txt
    pip-audit -r ingestion/requirements.txt
    pip-audit -r pb-proxy/requirements.txt
    pip-audit -r worker/requirements.txt
```

1. `run:` uses the default `bash -e`, so the shell exits at the first non-zero
   command. `pip-audit` exits 1 when it finds something, so a finding in the
   second file skips the remaining two.
2. `continue-on-error: true` then hides that the step failed at all, so the check
   is green and nobody looks.

The effect is worse than no audit for those two files, because the job actively
signals that they were checked.

## Suggested fix

Audit every file and report a single collected status:

```yaml
run: |
  status=0
  for svc in mcp-server ingestion pb-proxy worker; do
    echo "::group::pip-audit $svc"
    pip-audit -r "$svc/requirements.txt" || status=1
    echo "::endgroup::"
  done
  exit $status
```

## Open question

Whether `continue-on-error: true` should stay is a separate policy decision. It
keeps a newly published CVE in a transitive dependency from blocking every PR,
which is defensible — but then the finding needs to be visible somewhere other
than a collapsed log section, e.g. written to `$GITHUB_STEP_SUMMARY`. Right now a
real finding is indistinguishable from a clean run at a glance.

Worth fixing before the audit is relied on: the finding it currently hides is
tracked separately.

--- a/.github/workflows/pr-validate.yml
+++ b/.github/workflows/pr-validate.yml
@@ -106,16 +106,68 @@ jobs:
         with:
           python-version: "3.12"
 
+      # Constrained rather than pinned inline: requirements-dev.txt already
+      # carries the pins, and an unpinned `pip install` of the scanner is the
+      # exact supply-chain hole this job exists to find — see CLAUDE.md
+      # "Dependency Pinning".
       - name: Install scanning tools
-        run: pip install pip-audit bandit
+        run: pip install --constraint requirements-dev.txt pip-audit bandit
 
+      # pip-audit exits 1 when it finds something. The four invocations used to
+      # be four plain lines under the default `bash -e`, so the first finding
+      # aborted the step and the files after it were never audited at all —
+      # while `continue-on-error` reported the job green. That is worse than no
+      # audit, because the step claims coverage it did not deliver (#257).
+      #
+      # Every file is audited now and the statuses are collected. Coverage comes
+      # from `git ls-files` instead of a hand-kept list, so a newly added
+      # requirements file cannot be silently left out by being forgotten here —
+      # the same gap `.github/dependabot.yml` guards against for updates.
+      # `--strict` makes a dependency that could not be resolved a failure
+      # rather than a skipped line in the middle of a passing run.
+      #
+      # `continue-on-error` stays: a CVE published in a transitive dependency
+      # should not block every unrelated PR. It no longer hides the finding —
+      # the result lands in the job summary and raises an annotation, so a
+      # finding is distinguishable from a clean run at a glance.
       - name: Dependency audit (pip-audit)
         continue-on-error: true
         run: |
-          pip-audit -r mcp-server/requirements.txt
-          pip-audit -r ingestion/requirements.txt
-          pip-audit -r pb-proxy/requirements.txt
-          pip-audit -r worker/requirements.txt
+          status=0
+          rows=$(mktemp)
+          details=$(mktemp)
+
+          while IFS= read -r req; do
+            log=$(mktemp)
+            echo "::group::pip-audit $req"
+            if pip-audit --strict -r "$req" >"$log" 2>&1; then
+              printf '| `%s` | no known vulnerabilities |\n' "$req" >>"$rows"
+            else
+              status=1
+              printf '| `%s` | :warning: **findings** |\n' "$req" >>"$rows"
+              {
+                printf '\n<details><summary><code>%s</code></summary>\n\n```\n' "$req"
+                cat "$log"
+                printf '```\n\n</details>\n'
+              } >>"$details"
+            fi
+            cat "$log"
+            echo "::endgroup::"
+          done < <(git ls-files '*requirements*.txt')
+
+          {
+            echo "## Dependency audit (pip-audit)"
+            echo
+            echo "| requirements file | result |"
+            echo "| --- | --- |"
+            cat "$rows"
+            cat "$details"
+          } >>"$GITHUB_STEP_SUMMARY"
+
+          if [ "$status" -ne 0 ]; then
+            echo "::warning title=Dependency audit::pip-audit reported findings — see the job summary for the affected requirements files."
+          fi
+          exit "$status"
 
       - name: Static security scan (bandit)
         continue-on-error: true

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,31 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Fixed
+
+- **The dependency audit skipped files and still reported success** — the
+  `security-scan` job ran four `pip-audit` invocations as four plain lines in a
+  `run:` block. GitHub runs those under `bash -e`, and `pip-audit` exits 1 when
+  it finds something, so a finding in the second file ended the step before the
+  third and fourth were ever audited. `continue-on-error: true` then swallowed
+  the failure and the check went green. Two requirement files were therefore
+  unaudited while the job actively signalled that they had been checked — worse
+  than having no audit at all, because nobody goes looking. The step now audits
+  each file, collects the statuses instead of aborting, and exits with the
+  collected result. The file list comes from `git ls-files '*requirements*.txt'`
+  rather than being spelled out, which also closes the four files that were
+  never in the list (`reranker`, `demo`, `requirements-dev`, the Office 365
+  adapter), and means a requirements file added later cannot be left out by
+  being forgotten. `--strict` turns a dependency that could not be resolved into
+  a reported failure rather than a skipped line inside a passing run.
+  `continue-on-error` stays — a CVE published in a transitive dependency should
+  not block every unrelated PR — but a finding is no longer indistinguishable
+  from a clean run: the per-file result is written to the job summary with the
+  audit output inlined, and a workflow annotation points at it. The scanners
+  themselves are installed with `--constraint requirements-dev.txt` instead of
+  unpinned, since an unpinned `pip install` in CI is the supply-chain hole this
+  job exists to find. (#257)
+
 ## [0.12.1] - 2026-08-17
 
 ### Fixed

